docs/tutorial/reduction.rst

```diff
@@ -0,0 +1,720 @@
+    {
+    -    extern __shared__ T shared[];
+    +    __shared__ T shared[BlockSize];
+
+        ...
+
+        // Shared reduction
+    -    for (uint32_t i = blockDim.x / 2; i > WarpSize; i /= 2)
+    +    tmp::static_for<BlockSize / 2, tmp::greater_than<WarpSize>, tmp::divide<2>>([&]<int I>()
+        {
+    -        if (tid < i)
+    -            shared[tid] = op(shared[tid], shared[tid + i]);
+    +        if (tid < I)
+    +            shared[tid] = op(shared[tid], shared[tid + I]);
+            __syncthreads();
+        }
+    +    );
+
+Introducing yet another template argument for the kernel and moving from ``for`` to ``tmp::static_for`` leads to the following two notable improvements:
+
+- Introducing new attribute ``__launch_bounds__(BlockSize)`` to the kernel instructs the compiler that the kernel will only be launched using the designated block size. This implies that the launches of differing block sizes will fail. This allows the optimizer to enroll the ``blockDim.x`` variable in constant folding as well as get information about register usage.
+- Turning the block size into a compile-time constant allows you to statically allocate the shared memory.
+
+Communicate using warp-collective functions
+-------------------------------------------
+
+Shared memory provides a fast communication path within a block, however when performing reduction within the last warp, you can use faster means of communication, which is warp-collective or cross-lane functions. Instead of using the hardware-backed shared memory, you can directly copy between the local memory (registers) of each lane in a warp. This can be achieve using the shuffle functions.
+
+See how to use ``__shfl_down()``, which is one of the most restrictive but also the most structured communication schemes.
+
+.. code-block:: C++
+
+    // Warp reduction
+    if (tid < WarpSize)
+    {
+        T res = op(shared[tid], shared[tid + WarpSize]);
+        tmp::static_for<WarpSize / 2, tmp::not_equal<0>, tmp::divide<2>>([&]<int Delta>()
+        {
+            res = op(res, __shfl_down(res, Delta));
+        });
+
+        // Write result from shared to back buffer
+        if (tid == 0)
+            back[bid] = res;
+    }
+
+Using warp-collective functions for communication requires the control flow to be uniform across warps, as the name warp-collective implies. Therefore, you can see that the thread id is being checked outside the loop, but the result is written inside due to variable scoping.
+
+Prefer warp communication over shared
+-------------------------------------
+
+As mentioned in the previous step, communication between local memory is faster than shared memory. Instead of relying on the local memory only at the end of the tree-like reduction, it is a better approach to turn the tree reduction inside out and perform multiple warp reductions in parallel on all active threads, thus communicating only their partial results through the shared memory.
```

Comment:
```suggestion
As mentioned in the previous step, communication between local memory is faster than shared memory. Instead of relying on the local memory only at the end of the tree-like reduction, a better approach is to turn the tree reduction inside out and perform multiple warp reductions in parallel on all active threads, thus communicating only their partial results through the shared memory.
```